// @aws/dynamodb-expressions/build/MathematicalExpression.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var AttributePath_1 = require("./AttributePath");
var MATHEMATICAL_EXPRESSION_TAG = 'AmazonDynamoDbMathematicalExpression';
var EXPECTED_TOSTRING = "[object " + MATHEMATICAL_EXPRESSION_TAG + "]";
/**
 * An object representing a DynamoDB function expression.
 */
var MathematicalExpression = /** @class */ (function () {
    function MathematicalExpression(lhs, operator, rhs) {
        this.lhs = lhs;
        this.operator = operator;
        this.rhs = rhs;
        this[Symbol.toStringTag] = MATHEMATICAL_EXPRESSION_TAG;
    }
    MathematicalExpression.prototype.serialize = function (attributes) {
        var safeArgs = [this.lhs, this.rhs].map(function (arg) { return AttributePath_1.AttributePath.isAttributePath(arg) || typeof arg === 'string'
            ? attributes.addName(arg)
            : attributes.addValue(arg); });
        return safeArgs[0] + " " + this.operator + " " + safeArgs[1];
    };
    /**
     * Evaluate whether the provided value is a MathematicalExpression object.
     */
    MathematicalExpression.isMathematicalExpression = function (arg) {
        return arg instanceof MathematicalExpression
            || Object.prototype.toString.call(arg) === EXPECTED_TOSTRING;
    };
    return MathematicalExpression;
}());
exports.MathematicalExpression = MathematicalExpression;
